File: solution/1287. Element Appearing More Than 25% In Sorted Array/1287_03.cpp
class Solution {
public:
    int findSpecialInteger(vector<int>& arr) {
        const int n = arr.size();
        vector<int> A { arr[n/4], arr[n/2], arr[n*3/4] }; // 斬開4份, 只有這3個數有可能是答案
        for (int x : A) {
            auto l = lower_bound(arr.begin(), arr.end(), x);
            auto r = upper_bound(arr.begin(), arr.end(), x);
            if (distance(l, r) > n/4)
                return x;
        }
        return -1;
    }
};

File: solution/1287. Element Appearing More Than 25% In Sorted Array/1287_03.py
class Solution:
    def findSpecialInteger(self, arr: List[int]) -> int:
        n = len(arr)
        
        def first_occurence(x):
            l, r = 0, n
            while l < r:
                m = (l + r) // 2
                if arr[m] < x:
                    l = m + 1
                else:
                    r = m
            return l
        
        def last_occurence(x):
            l, r = 0, n
            while l < r:
                m = (l + r) // 2
                if arr[m] <= x:
                    l = m + 1
                else:
                    r = m
            return l - 1
        
        for x in [arr[n//4], arr[n//2], arr[n*3//4]]:
            l, r = first_occurence(x), last_occurence(x)
            if r - l + 1 > n/4:
                return x
        return -1
